plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java

```diff
@@ -0,0 +1,680 @@
+    public static final String IGNORE_MISSING = "ignore_missing";
+    public static final String FUNCTION_NAME = "function_name";
+    public static final String FULL_RESPONSE_PATH = "full_response_path";
+    public static final String MODEL_INPUT = "model_input";
+    public static final String DEFAULT_MODEl_INPUT = "{ \"parameters\": ${ml_inference.parameters} }";
+    // At default, ml inference processor allows maximum 10 prediction tasks running in parallel
+    // it can be overwritten using max_prediction_tasks when creating processor
+    public static final int DEFAULT_MAX_PREDICTION_TASKS = 10;
+    public static final String DEFAULT_OUTPUT_FIELD_NAME = "inference_results";
+
+    protected MLInferenceSearchResponseProcessor(
+        String modelId,
+        List<Map<String, String>> inputMaps,
+        List<Map<String, String>> outputMaps,
+        Map<String, String> modelConfigMaps,
+        int maxPredictionTask,
+        String tag,
+        String description,
+        boolean ignoreMissing,
+        String functionName,
+        boolean fullResponsePath,
+        boolean ignoreFailure,
+        boolean override,
+        String modelInput,
+        Client client,
+        NamedXContentRegistry xContentRegistry
+    ) {
+        super(tag, description, ignoreFailure);
+        this.inferenceProcessorAttributes = new InferenceProcessorAttributes(
+            modelId,
+            inputMaps,
+            outputMaps,
+            modelConfigMaps,
+            maxPredictionTask
+        );
+        this.ignoreMissing = ignoreMissing;
+        this.functionName = functionName;
+        this.fullResponsePath = fullResponsePath;
+        this.ignoreFailure = ignoreFailure;
+        this.override = override;
+        this.modelInput = modelInput;
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    public SearchResponse processResponse(SearchRequest request, SearchResponse response) throws Exception {
+        throw new RuntimeException("ML inference search response processor make asynchronous calls and does not call processRequest");
+    }
+
+    /**
+     * Processes the search response asynchronously by rewriting the documents with the inference results.
+     *
+     * @param request         the search request
+     * @param response        the search response
+     * @param responseContext the pipeline processing context
+     * @param responseListener the listener to be notified when the response is processed
+     */
+    @Override
+    public void processResponseAsync(
+        SearchRequest request,
+        SearchResponse response,
+        PipelineProcessingContext responseContext,
+        ActionListener<SearchResponse> responseListener
+    ) {
+        try {
+            SearchHit[] hits = response.getHits().getHits();
+            // skip processing when there is no hit
+            if (hits.length == 0) {
```

Comment:
Check if hits is null first? According to my below comment, maybe 
`if (response.getHits().getHits() == null || response.getHits().getHits().length == 0)`, 
then
`rewriteResponseDocuments(response, responseListener)`  ?

Reply:
when the hits is null but not empty, it should throw exception and we catch it by ignoreFailure flag, if ignoreFailure is false, it will throw exception accordingly. A valid search response with not matching any document is like this 

An empty hit when hits.length == 0, we skip processing the ml inference here. 
```
{
  "took": 45,
  "timed_out": false,
  "_shards": {
    "total": 404,
    "successful": 404,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 0,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  }
}
```

Reply:
>if ignoreFailure is false, it will throw exception accordingly.

So in this case we will show "Null pointer exception" to the customer? Does this sound reasonable to customer to get a null pointer exception and having no clue what exactly happened?